firmware/version_data.py

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python
+
+import subprocess
+import os
+import tempfile
+import filecmp
+import shutil
+
+commit = subprocess.check_output(['git', 'log', '--format="%H"', '-n', '1'])
+print("Showing commit variable")
+print(commit[1:-2])
+
+branch = subprocess.check_output(['git', 'symbolic-ref', '--short', 'HEAD'])
+print("Showing branch variable")
+print(branch[:-1])
+
+describe = subprocess.check_output(['git', 'describe', '--dirty'])
+print("Showing describe variable")
+print(describe[:-1])
+
+status = subprocess.check_output(['git', 'status', '--short'])
+length = status.count(b'\n')
+print(status)
+
+print("Showing uplatform")
+if "PLATFORM" in os.environ:
+    platform = os.environ['PLATFORM']
+    print(platform.upper())
+else:
+    platform = ""
+
+print("Showing utarget")
+if "TARGET" in os.environ:
+    target = os.environ['TARGET']
+    print(target.upper())
+else:
+    target = ""
+
+temp_h = tempfile.NamedTemporaryFile(suffix='.h', delete='true')
+print("Showing temp file .h")
+print(temp_h.name)
+temp_h.write(b"""\
+#ifndef __VERSION_DATA_H
+#define __VERSION_DATA_H
+extern const char* board;
+extern const char* target;
+extern const char* git_commit;
+extern const char* git_branch;
+extern const char* git_describe;
+extern const char* git_status;
+#endif  // __VERSION_DATA_H""")
+temp_h.seek(0)
+
+
+temp_c = tempfile.NamedTemporaryFile(suffix='.c', delete='true')
+print("Showing temp file .c")
+print(temp_c.name)
+
+temp_c.write(b"""\
+
+#ifndef PLATFORM_%s
+#error \"Version mismatch - PLATFORM_%s not defined!\"
```

Comment:
You don't need to escape the quotes inside triple quote.

You should also look at the [`string.format`](https://docs.python.org/3.4/library/string.html#string-formatting) here rather than string interpolation. I think it'll reduce the amount of duplication.